### ISSUE TITLE
Add actual documentation.

### ISSUE THREAD

**Issue by jantic** [jantic/DeOldify#13]
Standard Python class/function documentation, basically.

**Comment by avaika** [comment]
Please also include detailed step-by-step guide on how to apply the project on new images with ipython / etc. Not every potential user is familiar with that and it might be quite tricky to figure it out without a hint.

**Comment by jantic** [comment]
> Please also include detailed step-by-step guide on how to apply the project on new images with ipython / etc. Not every potential user is familiar with that and it might be quite tricky to figure it out without a hint.

I'll keep that in mind.  For now there's jupyter notebooks that already run with examples that I've pointed the way to.  I have to prioritize efforts here so for now this will be on the back burner.

**Comment by RihamHazem** [comment]
Can you just explain the architecture used in the generator? and how you combined the resnet34 with U-Net as I didn't manage to understand it from the code.

Thanks in advance.

**Comment by jantic** [comment]
Well...first of all, based on the way you're posing the question "how you combined the resnet34 with U-Net", that's indicating to me that there might be a misunderstanding of what U-Net itself is.  U-Net is half image recognition model (such as resnet), acting as an "encoder".  In the case of DeOldify, that resnet34 is already pretrained for free, meaning we get to focus mostly on training the other half- the decoder.  The decoder just takes what the feature detecting encoder "encoded", and generates an image from that by making decision on the features indicated by the encoder.  A big part of the magic of U-Net and how it achieves particularly high quality images is the fact that that information loss is minimized by extra "cross-connections" from encoder layers to the equivalent decoder layers.  

i.e it's just this diagram, except replace the left half with a resnet.  

https://pics.spark-in.me/upload/98bb30833b833a1ef9199feca3b9ed5c.png

Basically the Generator is a standard U-Net (as described above) plus spectral normalization added to the layers of decoder end.

**Comment by jantic** [comment]
I guess I should clarify:  Not standard U-Net as in the original implementation but standard U-Net as in the basic thrust of what a U-Net does, if that makes sense.  The configuration of number of filters and other such details are different, for sure, but that shouldn't distract from the big picture.

**Comment by RihamHazem** [comment]
Thanks a lot for this rapid response and the clear explanation, it really helped me a lot in understanding the network and thanks for sharing such awesome work.
I've another question concerning the GANTrainSchedule. Does its purpose mainly is to tune the model's hyper-parameters by trying different values for them? and if so is it available to know what's the best value for those hyper-parameters?

**Comment by jantic** [comment]
Thanks!

Unfortunately it's not tuning any hyperparameters for you.  In fact that's precisely the part of this work that's such a pain- there's really a narrow set that actually works with learning rates etc.  There's new work I've been doing that will be revealed next week that alleviates all this thankfully.  

So all the GANTrainSchedule really does for you is it allows you to pass an ordered list of configurations for training.  You're configuring things like batch size, learning rate, etc.  The main motivation of this was to cut down on code, but I've abandoned this approach in the latest iteration.  It doesn't really seem to actually help much in practice (code should evolve like that!)

**Comment by jantic** [comment]
Oh and the best values for the hyper-parameters, as far as I can tell, are the ones chosen already in the notebooks.  Those came out of a lot of trial and error.

**Comment by RihamHazem** [comment]
Okay I've understood the functionality of GANTrainerSchedule but you're saying that you found out that this approach in training doesn't really help in improving the results or I misunderstand you 😅 sorry for asking so much questions.

**Comment by jantic** [comment]
@RihamHazem Sorry, just saw this now.  I'm saying that GANTrainerSchedule was just burdensome and unnecessary code that got in the way overall, and that it was better to just implement the steps it executed as individual cells in Jupyter. i.e. it's just a refactoring change I'm talking about.  Though, separately, training actually did change quite dramatically in the latest DeOldify.  But these are two separate things.

**Comment by jantic** [comment]
There's sufficient documentation for my liking now.  I still am a firm believer in "self-documenting code" and don't think every line of code should have comment but should be clear enough to stand on its own in understandability.  If I'm failing on that...then I'll fix the code!

**Comment by jorgealbertogomezgomez77** [comment]
The model can be used with images provided from the computer and not from the internet?

**Comment by jantic** [comment]
@jorgealbertogomezgomez77  Yes, if you install DeOldify locally you can colorize images and video from your local drive.  In the ImageColorizer.ipynb notebook, for example, you just set the source_url to None and put the local file path in for source_path .  Admittedly the current implementations in the notebooks are a bit goofy.

![image](https://user-images.githubusercontent.com/179759/90066102-27470580-dca2-11ea-9532-3d47017c87f6.png)
